File: rust/src/n0278_first_bad_version.rs
pub fn first_bad_version<T: Fn(i32) -> bool>(n: i32, is_bad_version: T) -> i32 {
    let mut start = 0;
    let mut end = n;
    while start < end {
        let mid = start + (end - start) / 2;
        if is_bad_version(mid) {
            end = mid;
        } else {
            start = mid + 1;
        }
    }
    end
}

File: typescript/src/n0278_first_bad_version.ts
export function firstBadVersion(
  n: number,
  isBadVersion: (version: number) => boolean,
): number {
  let start = 0;
  let end = n;
  while (start < end) {
    const mid = Math.floor((start + end) / 2);
    if (isBadVersion(mid)) {
      end = mid;
    } else {
      start = mid + 1;
    }
  }
  return end;
}
